### PR TITLE
Phase 1 state machines

### PR DESCRIPTION
**Pull request by garthwatney**
This pull request adds the Phase 1 of state machines to the FPP language.  In phase 1 the state machine behavior is specified outside FPP but the state machine definition names and instance names are specified in FPP and the appropriate support code is generated in the component base class.  There are also some incidental cleanup like the key words (ask Rob)

### CODE REVIEW

**Comment by bocchino** [comment]
We held an in-person review on 8/20/24. At that review we agreed on the following changes:
* There should be one signal-sending function for each state machine instance, instead of one function for all instances.
* The signal-sending function should implicitly fill in the state machine instance ID and should have two arguments: a signal ID enumeration ID and SignalBuffer, which is a stack-allocated buffer. The SignalBuffer should just contain data bytes.
* The size of the data stored in the SignalBuffer should not be hard-coded. It should be configurable as part of the F Prime configuration.
* The pure virtual functions called when doing an action should have three arguments: state machine instance ID, signal ID, and signal buffer.

**Comment by garthwatney** [comment]
All issues from the 8/20 review should be fixed.

**Comment by bocchino** [comment]
Great! I confirmed that the `check-cpp` tests pass.

As discussed, let's exercise a few more capabilities in the state machines in the unit tests.

File: compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp [hunk 1]
```diff
@@ -13,6 +13,12 @@ module Fw {
     NO_ROOM_LEFT  @< No room left in the buffer to serialize data
   }
 
+  struct SMSignals {
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp:16]: Do we still need this type?

File: compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp [hunk 2]
```diff
@@ -71,4 +77,10 @@ module Fw {
       FAILURE @< Representing failure
       SUCCESS @< Representing success
   }
+
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp:80]: Why is this type being added?

**Comment by bocchino** [comment]
> I confirmed that the check-cpp tests pass.

Scratch that, I forgot to pull the latest. It looks like the STARS-generated code needs to be updated in this repo in order for the `fpp-check` compilation to work.

**Comment by bocchino** [comment]
After the updates, I ran fpp-check against this F Prime branch: https://github.com/garthwatney/fprime/commit/08d8135b341daf5531a0ed1c1c06bad6bfe636f5. I'm seeing this compilation error in fpp-to-cpp/test/component/impl:
```
In file included from ActiveStateMachines.cpp:7:
./ActiveStateMachines.hpp:40:17: error: use of undeclared identifier 'S2_Interface'
          const S2_Interface::S2_Signals signal, //!< The state machine signal
                ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

File: compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp (hunk 2 above)
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp:80]: This type is coming in from F Prime.

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+// ======================================================================
+// \title  ActiveStateMachines_S1.cpp
+// \author Auto-generated
+// \brief  cpp file for state machine ActiveStateMachines_S1
+//
+// ======================================================================            
+    
+#include "stdio.h"
+#include "assert.h"
+#include "Fw/Types/SMSignalsSerializableAc.hpp"
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp:11]: We need to remove this header include. `SMSignals` is no longer used.

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+// \title  ActiveStateMachines_S1.h
+// \author Auto-generated
+// \brief  header file for state machine ActiveStateMachines_S1
+//
+// ======================================================================
+           
+#ifndef ACTIVESTATEMACHINES_S1_H_
+#define ACTIVESTATEMACHINES_S1_H_
+                                
+#include <Fw/SMSignal/SMSignalBuffer.hpp>
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp:12]: For consistency with the rest of F Prime, I think the path should be `Fw/Sm/SmSignalBuffer.hpp`. Compare `Fw/Com/ComPacket.hpp`, `Fw/Dp/DpContainer.hpp`, etc.

File: compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp (hunk 1 above)
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/fprime/Fw/Types/Types.fpp]: I removed this type and fixed a related bug in the code gen (a use of the type was still hanging around and needed to be replaced).

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp (hunk 3 above)
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp:11]: I removed this.

Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp]: I fixed this. I also removed the dependencies on stdio and assert. These are not allowed in F Prime flight code. @garthwatney Please update the STARS code generator to match.

**Comment by bocchino** [comment]
`fpp-check` is now passing.

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
+
+// ======================================================================
+// \title  ActiveStateMachines_S1.cpp
+// \author Auto-generated
+// \brief  cpp file for state machine ActiveStateMachines_S1
+//
+// ======================================================================            
+
+#include "ActiveStateMachines_S1.hpp"
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp:9]: @garthwatney Can you make sure that STARS generates these header includes? Before I saw dependencies on C stdio and C assert. We shouldn't have those dependencies in flight code.

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            
+            switch (signal) {
+
+                default:
+                    break;
+            }
+            break;
+    
+        default:
+            FW_ASSERT(0);
```
Comment (bocchino) [inline compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp:98]: @garthwatney Can you make sure that STARS generates FW_ASSERT here? Before I saw a C-style assert, which we should not use in F Prime flight code.

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp (hunk 4 above)
Comment (garthwatney) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp:12]: Done

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp (hunk 6 above)
Comment (garthwatney) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp]: Done

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp (hunk 5 above)
Comment (garthwatney) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.cpp:10]: Done

File: compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp (hunk 4 above)
Comment (bocchino) [reply compiler/tools/fpp-to-cpp/test/component/ActiveStateMachines_S1.hpp]: Looks good!

**Review by bocchino — APPROVED** [review]
Looks good!

**Review by LeStarch — APPROVED** [review]

**Pull request merged**